### PR TITLE
1.22 feature blog for minReadySeconds in StatefulSets

### PR DESCRIPTION
**Pull request by jlbutler**
1.22 feature blog post for "Add minReadySeconds to Statefulsets" (https://github.com/kubernetes/enhancements/issues/2599).

Target publish date is 08/27/21. [edit: updated publish date]

cc: @ravisantoshgudimetla

### CODE REVIEW

File: content/en/blog/_posts/2021-08-16-minreadysecond-statefulsets.md
Comment (sftim) [inline content/en/blog/_posts/2021-08-16-minreadysecond-statefulsets.md:14]: Q. Does “today” describe the situation before the v1.22 release, or after it?

File: content/en/blog/_posts/2021-08-16-minreadysecond-statefulsets.md
Comment (sftim) [inline content/en/blog/_posts/2021-08-16-minreadysecond-statefulsets.md:3]: Is there a typical format for the titles of post-release blog articles? If so, let's follow that.

File: content/en/blog/_posts/2021-08-16-minreadysecond-statefulsets.md
Comment (ravisantoshgudimetla) [reply content/en/blog/_posts/2021-08-16-minreadysecond-statefulsets.md:14]: It is before v1.22

File: content/en/blog/_posts/2021-08-16-minreadysecond-statefulsets.md
Comment (ravisantoshgudimetla) [reply content/en/blog/_posts/2021-08-16-minreadysecond-statefulsets.md:3]: I am not sure of that. Where do we have style guide?

File: content/en/blog/_posts/2021-08-16-minreadysecond-statefulsets.md
Comment (sftim) [reply content/en/blog/_posts/2021-08-16-minreadysecond-statefulsets.md:14]: As this article goes out after the release, let's reword to make that clear.

File: content/en/blog/_posts/2021-08-16-minreadysecond-statefulsets.md
Comment (ravisantoshgudimetla) [reply content/en/blog/_posts/2021-08-16-minreadysecond-statefulsets.md:3]: @jlbutler ^?

File: content/en/blog/_posts/2021-08-16-minreadysecond-statefulsets.md
Comment (sftim) [inline content/en/blog/_posts/2021-08-16-minreadysecond-statefulsets.md:43]:
```suggestion
- Read the KEP: [minReadySeconds for StatefulSets](https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/2599-minreadyseconds-for-statefulsets#readme)
- Read the documentation: [Minimum ready seconds](/docs/concepts/workloads/controllers/statefulset/#minimum-ready-seconds) for StatefulSet
- Review the [API definition](/docs/reference/kubernetes-api/workload-resources/stateful-set-v1/) for StatefulSet
```

File: content/en/blog/_posts/2021-08-16-minreadysecond-statefulsets.md
Comment (sftim) [inline content/en/blog/_posts/2021-08-16-minreadysecond-statefulsets.md:47]:
```suggestion
Please reach out to us in the [#sig-apps](https://kubernetes.slack.com/archives/C18NZM5K9) channel on Slack (visit https://slack.k8s.io/ for an invitation if you need one), or on the SIG Apps mailing list: kubernetes-sig-apps@googlegroups.com
```

File: content/en/blog/_posts/2021-08-16-minreadysecond-statefulsets.md
Comment (sftim) [inline content/en/blog/_posts/2021-08-16-minreadysecond-statefulsets.md:25]: tiny nit:
```suggestion
The statefulset controller watches for both `StatefulSets` and the `Pods` associated with them. When the feature gate associated with this feature is enabled, the statefulset controller identifies how long a particular `Pod` associated with a `StatefulSet` has been in the `Running` state.

If this value is greater than or equal to the time specified by the end user in `.spec.minReadySeconds` field, the statefulset controller updates a field called `availableReplicas` in the `StatefulSet`'s status subresource to include this `Pod`. The `.status.availableReplicas` field for a `StatefulSet` is an integer field that tracks the number of pods that are `Available`.
```

**Review by sftim — COMMENTED** [review]
A bit more feedback.

Please do update the article to frame it as being published after the v1.22 release. That's the most important change that needs making. The other details are less essential.

File: content/en/blog/_posts/2021-08-16-minreadysecond-statefulsets.md
Comment (jlbutler) [reply content/en/blog/_posts/2021-08-16-minreadysecond-statefulsets.md:3]: Apologies I thought I'd replied to this. As I mentioned last week, I also think this could use a more reader-friendly title. 

I don't know of any specific guideline but they do tend to be more descriptive of functionality, rather than calling out details.

File: content/en/blog/_posts/2021-08-16-minreadysecond-statefulsets.md
Comment (sftim) [inline content/en/blog/_posts/2021-08-16-minreadysecond-statefulsets.md:4]: It's handy to update the path to the Markdown file to match. Hugo doesn't care but us humans can get confused.

**Comment by jlbutler** [comment]
@ravisantoshgudimetla can you reply to review when you have a few cycles? thanks!

**Comment by sftim** [comment]
@jlbutler if you're willing to make the changes needed, feel free. That'll let us review.

If you want to get signoff on those changes from @ravisantoshgudimetla before we publish, add a /hold so we don't accidentally put it live without that signoff.

**Comment by ravisantoshgudimetla** [comment]
I'll get to this tomorrow. Thank you for the reviews @sftim

**Review by ravisantoshgudimetla — COMMENTED** [review]
Made the changes locally. Will push the changes to remote branch once I have access to it. The earlier collaboration invite is expired now :(

**Review by ravisantoshgudimetla — COMMENTED** [review]
@sftim - Made the changes you suggested modulo the title formatting. I am not sure which format I should use. Let me know if the current title looks good. TY

**Comment by ravisantoshgudimetla** [comment]
cc @soltysh

File: content/en/blog/_posts/2021-08-16-minreadysecond-statefulsets.md
Comment (simonpasquier) [inline content/en/blog/_posts/2021-08-16-minreadysecond-statefulsets.md:18]:
```suggestion
So, in general, the notion of `Availability` in `StatefulSets` is pretty useful and this feature helps in solving the above problems. This is a feature that already exists for `Deployments` and `DaemonSets` and we now have them for `StatefulSets` too to give users consistent workload experience.
```

Comment (ravisantoshgudimetla) [reply content/en/blog/_posts/2021-08-16-minreadysecond-statefulsets.md]: Thank you @simonpasquier for the review

**Comment by sftim** [comment]
I think the title for this article is OK and we should focus on getting this merged (we can update the title after merging if we think of a new one that we really like).

**Comment by sftim** [comment]
@ravisantoshgudimetla / @soltysh - between the two of you I'm happy that you can review each other's contributions here. Feel free to add /lgtm when you're happy that this can go out.

LGTM - for writing style etc

/hold
@jlbutler are you happy to revise this to show the new publication date?

/approve

**Comment by ravisantoshgudimetla** [comment]
> @ravisantoshgudimetla / @soltysh - between the two of you I'm happy that you can review each other's contributions here. Feel free to add /lgtm when you're happy that this can go out.

@soltysh has already reviewed it. We worked with @jlbutler before opening this PR

File: content/en/blog/_posts/2021-08-16-minreadysecond-statefulsets.md
Comment (chrisnegus) [inline content/en/blog/_posts/2021-08-16-minreadysecond-statefulsets.md:14]:
```suggestion
Prior to Kubernetes 1.22 release, once a `StatefulSet` `Pod` is in the `Ready` state it is considered `Available` to receive traffic. For some of the `StatefulSet` workloads, that may not be the case. For example, a workload like Prometheus, with multiple instances of Alertmanager, should be considered `Available` only when Alertmanager's state transfer is complete, not when the `Pod` is in `Ready` state. Since `minReadySeconds` adds buffer, the state transfer may be complete before the `Pod` becomes `Available`. While this is not a foolproof way of identifying if the state transfer is complete or not, it gives a way to the end user to express their intention of waiting for some time before the `Pod` is considered `Available` and ready to serve requests. 
```

File: content/en/blog/_posts/2021-08-16-minreadysecond-statefulsets.md
Comment (chrisnegus) [inline content/en/blog/_posts/2021-08-16-minreadysecond-statefulsets.md:16]:
```suggestion
Another case where `minReadySeconds` helps is when using `LoadBalancer` `Services` with cloud providers. Since `minReadySeconds` adds latency after a `Pod` is `Ready`, it provides buffer time to prevent killing pods in rotation before new pods show up. Imagine a load balancer in an unhappy path, taking 10-15s to propagate. If you have 2 replicas, then you'd kill the second replica only after the first one is up. But in reality, the first replica cannot be seen because it is not yet ready to serve requests.
```

File: content/en/blog/_posts/2021-08-16-minreadysecond-statefulsets.md
Comment (chrisnegus) [inline content/en/blog/_posts/2021-08-16-minreadysecond-statefulsets.md:36]:
```suggestion
Specify a value for `minReadySeconds` for any StatefulSet and you can check if `Pods` are available or not by checking the `AvailableReplicas` field using:
```

**Review by chrisnegus — COMMENTED** [review]
Looks good! Just a couple of small wording suggestions.

**Comment by sftim** [comment]
I pushed a commit to update the publication date.

@ravisantoshgudimetla / @soltysh once you're happy with the wording - _and “right now, actually” is a valid answer_ - please use the /lgtm command to record that you are. If you're not able to LGTM (“looks good to me”), or aren't comfortable running that command, reply to say so and SIG Docs can sort something out.

If we get this merged and folks want to fix minor grammar or spelling nits, those can go in after the merge (and either before or after publication).

If there's no LGTM before the publication date and we miss it by more than say 18 hours, we should leave the hold in place and pick a new date.

**Comment by sftim** [comment]
Oh, I should also add: if this gets LGTM _before_ publication is due on the 27th, please do unhold it.

**Review by ravisantoshgudimetla — APPROVED** [review]
Thank you @jlbutler @sftim @chrisnegus for the reviews

/lgtm

**Comment by sftim** [comment]
/hold cancel
:tada:

**Pull request merged**
